src/Browser.php

```diff
@@ -784,7 +880,55 @@ private function requestMayBeStreaming($method, $url, array $headers = array(),
         }
 
         return $this->transaction->send(
-            new Request($method, $url, $headers, $body, $this->protocolVersion)
+            new Request($method, $url, $this->mergeHeaders($headers), $body, $this->protocolVersion)
         );
     }
+
+    /**
+     * @param array $headers
+     * @return array
+     */
+    private function mergeHeaders($headers)
+    {
+        foreach (array_keys($this->defaultHeaders) as $key) {
+            foreach (array_keys($headers) as $headerKey) {
+                if (strcasecmp($headerKey, $key) === 0) {
+                    unset($this->defaultHeaders[$key]);
+                }
+            }
+        }
+
+        return array_merge($this->defaultHeaders, $headers);
+    }
+
+    /**
+     * @param array $headers
+     * @param string $header
+     * @return string|null
+     */
+    private function findCaseInsensitivHeader($headers, $header)
+    {
+        $found = null;
+        /** @var string $key */
```

Comment:
I agree that this seems reasonable and I'd love this to be true, but the array `["1" => "first"]` uses `(int) 1` as key automatically.

Reply:
the function got inlined. Im not that sure if that was your intention but i now changed it to "string|int $key". I could also just remove the annotation